File: backend/db/repositories/file-competency-repository.ts
/**
 * File-Based Competency Repository
 * Reads/writes competencies from JSON files instead of database
 */

import { readJSONFile, writeJSONFile } from '../../utils/file-storage.ts';
import type { CompetencyData } from "@shared/types/api.types";

interface FileCompetency {
  id: string;
  name: string;
  description: string;
  category?: string;
  weight?: number;
}

const COMPETENCIES_FILE = 'competencies.json';
const INDUSTRY_COMPETENCIES_FILE = 'industry-competencies.json';

export class FileCompetencyRepository {
  /**
   * Find competency by ID
   */
  async findById(id: string): Promise<CompetencyData | null> {
    try {
      const competencies = readJSONFile<FileCompetency[]>(COMPETENCIES_FILE);
      const competency = competencies.find(c => c.id === id);

      if (!competency) {
        return null;
      }

      return this.mapToCompetencyData(competency);
    } catch (error) {
      console.error('Error finding competency by ID:', error);
      throw error;
    }
  }

  /**
   * Get all competencies
   */
  async findAll(): Promise<CompetencyData[]> {
    try {
      const competencies = readJSONFile<FileCompetency[]>(COMPETENCIES_FILE);
      return competencies.map(c => this.mapToCompetencyData(c));
    } catch (error) {
      console.error('Error fetching all competencies:', error);
      throw error;
    }
  }

  /**
   * Get competencies by industry
   */
  async findByIndustry(industry: string): Promise<CompetencyData[]> {
    try {
      const industryCompetencies = readJSONFile<any>(INDUSTRY_COMPETENCIES_FILE);
      const allCompetencies = readJSONFile<FileCompetency[]>(COMPETENCIES_FILE);

      // Get competency IDs for this industry
      let competencyIds: string[] = [];

      if (industryCompetencies[industry]) {
        // Handle different structures (could be direct array or nested object)
        const industryData = industryCompetencies[industry];

        // Try to find competencies in various structures
        if (industryData.default && Array.isArray(industryData.default.competencies)) {
          competencyIds = industryData.default.competencies;
        } else if (industryData.default && Array.isArray(industryData.default)) {
          competencyIds = industryData.default;
        } else {
          // Get from first subcategory
          const firstSubcat = Object.keys(industryData)[0];
          if (firstSubcat && industryData[firstSubcat]) {
            if (Array.isArray(industryData[firstSubcat].competencies)) {
              competencyIds = industryData[firstSubcat].competencies;
            } else if (Array.isArray(industryData[firstSubcat])) {
              competencyIds = industryData[firstSubcat];
            }
          }
        }
      }

      // Map IDs to full competency objects
      const competencies = competencyIds
        .map(id => allCompetencies.find(c => c.id === id))
        .filter((c): c is FileCompetency => c !== undefined);

      return competencies.map(c => this.mapToCompetencyData(c));
    } catch (error) {
      console.error('Error fetching competencies by industry:', error);
      throw error;
    }
  }

  /**
   * Get competencies by category
   */
  async findByCategory(category: string): Promise<CompetencyData[]> {
    try {
      const competencies = readJSONFile<FileCompetency[]>(COMPETENCIES_FILE);
      const filtered = competencies.filter(c => c.category === category);
      return filtered.map(c => this.mapToCompetencyData(c));
    } catch (error) {
      console.error('Error fetching competencies by category:', error);
      throw error;
    }
  }

  /**
   * Create competency
   */
  async create(competencyData: Omit<CompetencyData, 'id'>): Promise<CompetencyData> {
    try {
      const competencies = readJSONFile<FileCompetency[]>(COMPETENCIES_FILE);

      // Generate ID from name
      const id = competencyData.name.toLowerCase().replace(/\s+/g, '-');

      const newCompetency: FileCompetency = {
        id,
        name: competencyData.name,
        description: competencyData.description,
        category: competencyData.category,
        weight: competencyData.weight,
      };

      competencies.push(newCompetency);
      writeJSONFile(COMPETENCIES_FILE, competencies);

      return this.mapToCompetencyData(newCompetency);
    } catch (error) {
      console.error('Error creating competency:', error);
      throw error;
    }
  }

  /**
   * Update competency
   */
  async update(id: string, competencyData: Partial<CompetencyData>): Promise<CompetencyData> {
    try {
      const competencies = readJSONFile<FileCompetency[]>(COMPETENCIES_FILE);
      const index = competencies.findIndex(c => c.id === id);

      if (index === -1) {
        throw new Error('Competency not found');
      }

      // Update fields
      if (competencyData.name !== undefined) {
        competencies[index].name = competencyData.name;
      }
      if (competencyData.description !== undefined) {
        competencies[index].description = competencyData.description;
      }
      if (competencyData.category !== undefined) {
        competencies[index].category = competencyData.category;
      }
      if (competencyData.weight !== undefined) {
        competencies[index].weight = competencyData.weight;
      }

      writeJSONFile(COMPETENCIES_FILE, competencies);

      return this.mapToCompetencyData(competencies[index]);
    } catch (error) {
      console.error('Error updating competency:', error);
      throw error;
    }
  }

  /**
   * Delete competency
   */
  async delete(id: string): Promise<void> {
    try {
      const competencies = readJSONFile<FileCompetency[]>(COMPETENCIES_FILE);
      const filtered = competencies.filter(c => c.id !== id);

      if (filtered.length === competencies.length) {
        throw new Error('Competency not found');
      }

      writeJSONFile(COMPETENCIES_FILE, filtered);
    } catch (error) {
      console.error('Error deleting competency:', error);
      throw error;
    }
  }

  /**
   * Get all categories
   */
  async getAllCategories(): Promise<string[]> {
    try {
      const competencies = readJSONFile<FileCompetency[]>(COMPETENCIES_FILE);
      const categories = [...new Set(competencies.map(c => c.category).filter(Boolean))] as string[];
      return categories.sort();
    } catch (error) {
      console.error('Error fetching competency categories:', error);
      throw error;
    }
  }

  /**
   * Map file competency to API format
   */
  private mapToCompetencyData(competency: FileCompetency): CompetencyData {
    return {
      id: competency.id,
      name: competency.name,
      description: competency.description,
      category: competency.category || 'General',
      weight: competency.weight || 10,
    };
  }
}

// Export singleton instance
export const competencyRepository = new FileCompetencyRepository();

File: backend-python/app/repositories/file_competency_repository.py
"""
File-Based Competency Repository
Reads/writes competencies from JSON files instead of database
"""

import logging
from typing import Any
from uuid import uuid4

from app.utils.file_storage import read_json_file, write_json_file
from app.models.competency import CompetencyData, CreateCompetencyRequest, UpdateCompetencyRequest

logger = logging.getLogger(__name__)

COMPETENCIES_FILE = "competencies.json"


class FileCompetencyRepository:
    """File-based repository for competency operations."""

    def _load_competencies(self) -> list[dict[str, Any]]:
        """Load competencies from JSON file."""
        return read_json_file(COMPETENCIES_FILE) or []

    def _save_competencies(self, competencies: list[dict[str, Any]]) -> None:
        """Save competencies to JSON file."""
        write_json_file(COMPETENCIES_FILE, competencies)

    async def find_by_id(self, competency_id: str) -> CompetencyData | None:
        """Find competency by ID."""
        try:
            competencies = self._load_competencies()
            for comp in competencies:
                if comp.get("id") == competency_id:
                    return CompetencyData(**comp)
            return None
        except Exception as e:
            logger.error(f"Error finding competency by ID: {e}")
            raise

    async def find_all(self) -> list[CompetencyData]:
        """Get all competencies."""
        try:
            competencies = self._load_competencies()
            return [CompetencyData(**comp) for comp in competencies]
        except Exception as e:
            logger.error(f"Error fetching all competencies: {e}")
            raise

    async def find_by_industry(self, industry: str) -> list[CompetencyData]:
        """Find competencies by industry."""
        try:
            competencies = self._load_competencies()
            filtered = [
                comp for comp in competencies
                if comp.get("industry") == industry or comp.get("industry") is None
            ]
            return [CompetencyData(**comp) for comp in filtered]
        except Exception as e:
            logger.error(f"Error finding competencies by industry: {e}")
            raise

    async def find_by_category(self, category: str) -> list[CompetencyData]:
        """Find competencies by category."""
        try:
            competencies = self._load_competencies()
            filtered = [comp for comp in competencies if comp.get("category") == category]
            return [CompetencyData(**comp) for comp in filtered]
        except Exception as e:
            logger.error(f"Error finding competencies by category: {e}")
            raise

    async def create(self, competency_data: CreateCompetencyRequest | dict[str, Any]) -> CompetencyData:
        """Create a new competency."""
        try:
            if isinstance(competency_data, CreateCompetencyRequest):
                data = competency_data.model_dump()
            else:
                data = competency_data

            competencies = self._load_competencies()

            new_competency = {
                "id": str(uuid4()),
                **data,
            }
            competencies.append(new_competency)
            self._save_competencies(competencies)

            return CompetencyData(**new_competency)
        except Exception as e:
            logger.error(f"Error creating competency: {e}")
            raise

    async def update(self, competency_id: str, competency_data: UpdateCompetencyRequest | dict[str, Any]) -> CompetencyData:
        """Update a competency."""
        try:
            if isinstance(competency_data, UpdateCompetencyRequest):
                data = competency_data.model_dump(exclude_unset=True)
            else:
                data = {k: v for k, v in competency_data.items() if v is not None}

            competencies = self._load_competencies()
            for i, comp in enumerate(competencies):
                if comp.get("id") == competency_id:
                    competencies[i] = {**comp, **data}
                    self._save_competencies(competencies)
                    return CompetencyData(**competencies[i])

            raise ValueError("Competency not found")
        except Exception as e:
            logger.error(f"Error updating competency: {e}")
            raise

    async def delete(self, competency_id: str) -> None:
        """Delete a competency."""
        try:
            competencies = self._load_competencies()
            competencies = [comp for comp in competencies if comp.get("id") != competency_id]
            self._save_competencies(competencies)
        except Exception as e:
            logger.error(f"Error deleting competency: {e}")
            raise


# Singleton instance
file_competency_repository = FileCompetencyRepository()
